rmf_site_format/src/workcell.rs

```diff
@@ -289,6 +344,129 @@ impl Workcell {
+        let workcell = self;
+
+        let visuals_and_parents: Vec<(urdf_rs::Visual, _)> = workcell
+            .visuals
+            .iter()
+            .map(|v| {
+                let visual = &v.1.bundle;
+                let visual = urdf_rs::Visual {
+                    name: Some(visual.name.clone()),
+                    origin: visual.pose.into(),
+                    geometry: visual.geometry.clone().into(),
+                    material: None,
+                };
+                (visual, v.1.parent)
+            })
+            .collect();
+        let mut parent_to_visuals = HashMap::new();
+        visuals_and_parents.iter().for_each(|(visual, parent)| {
+            parent_to_visuals
+                .entry(*parent)
+                .or_insert_with(Vec::new)
+                .push(visual.clone());
+        });
+
+        let collisions_and_parents: Vec<(urdf_rs::Collision, _)> = workcell
+            .collisions
+            .iter()
+            .map(|c| {
+                let collision = &c.1.bundle;
+                let collision = urdf_rs::Collision {
+                    name: Some(collision.name.clone()),
+                    origin: collision.pose.into(),
+                    geometry: collision.geometry.clone().into(),
+                };
+                (collision, c.1.parent)
+            })
+            .collect();
+        let mut parent_to_collisions = HashMap::new();
+        collisions_and_parents
+            .iter()
+            .for_each(|(collision, parent)| {
+                parent_to_collisions
+                    .entry(*parent)
+                    .or_insert_with(Vec::new)
+                    .push(collision.clone());
+            });
+
+        let links = workcell
+            .frames
+            .iter()
+            .map(|f| {
+                let frame = &f.1.bundle;
+
+                let name = match &frame.name {
+                    Some(name) => name.0.clone(),
+                    None => format!("frame_{}", f.0),
+                };
+
+                let pose = if let Anchor::Pose3D(p) = frame.anchor {
+                    p.into()
+                } else {
+                    return Err(WorkcellToUrdfError::InvalidAnchorType(frame.anchor.clone()));
+                };
+                let inertial = urdf_rs::Inertial {
+                    origin: pose,
+                    inertia: {
+                        urdf_rs::Inertia {
+                            ixx: 0.0,
+                            ixy: 0.0,
+                            ixz: 0.0,
+                            iyy: 0.0,
+                            iyz: 0.0,
+                            izz: 0.0,
+                        }
+                    },
+                    mass: urdf_rs::Mass { value: 0.0 },
+                };
+
+                let parent = &f.1.parent;
+                let collision = parent_to_collisions
+                    .get(parent)
+                    .map(|collisions| collisions.clone())
```

Comment:
I think the hierarchy calculation here might be off.
This line (and the same afterwards for visuals), do a lookup in the visual / collision hash map for the parent which is fine.
However, the `parent` variable doesn't contain the link's id but the link's parent id, so I believe this line would assign, to the current link, the visuals / collisions that are children of the link's parent link.
The `parent` should be the frame's ID `f.0` rather than `&f.1.parent`, for readability can just unpack the tuple at the iteration level by changing:

```
.map(|f| {
```

Into

```
.map(|(frame_id, frame) {
```

With this change you should be able to also simplify the HashMap access to move values and remove the clone.
Since in every loop iteration the key you use to access the map is different, you can just move the visuals / collisions out of the map and avoid the cloning as such:

```
let visual = parent_to_visuals
    .remove(frame_id)
    .unwrap_or_default();
```

Reply:
I noticed that hierarchy is also off in the example in your first post, the visual has a parent id of `0` that corresponds to the ID of the root of the workcell, while in the exported `urdf` its parent is set to the `top_anchor` frame which has id 2.
Now there is a minor challenge in the fact that the root might not be a frame itself but that should be pretty easy to fix by adding a dummy link at the origin

Reply:
This should be fixed now in 6c5999c.